### PR TITLE
fix(core): st-mixin feature overall

### PR DESCRIPTION
**Pull request by idoros**
## fix
- [x] mixin selector within nested selector: `:is(.mix)` not included in mixin: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%284%21%27.20715is-C%2520%252F*%2520bug6Not%2520Acluded%2520828B%291.252-is-711%252F*28startB.Ato3-st-2A67%27%29%7Eselected%214%27%2903origA61%255Cn2mix3%2520%281994%27%252Fentry.st.css5%253Ais%257B.2%257D06%253A%25207C1%298A%25209%2520%2520AinB*%252F1C2%253B%2501CBA9876543210_)
- [ ] mixin root out of order: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28A%21%27.a0aB5.909B5.b0bB55%252F*7aC14aE6aG79C249E69G7bC34bE6bGJ*%252F5.Ito8-st-mixF9B%27%29%7Eselected%21A%27%290864%255DDItoD5%255Cn6origF7J%2540rule%257B.8E5HH9rootA%27%252FIdex.st.cssB%253B5%29C%257D%255BD%2520.entry__E%2520%28FI%253A%2520G%253B%29H%2520%2520IinJ%25205%2501JIHGFEDCBA9876540_)
- [ ] mixin root specificity modifies root specificity result: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%283%21%27.8721.a74a611%252F*mix%2520start*%252F1.into0-st-mix2%27%29%7Eselected%213%27%290%2520%281551%255Cn24863%27%252Findex.st.css4in%253A%25205%2520%25206%253B1%2970orig8root%2501876543210_)
- [ ] root mixin attach `:global()` without the mixin anchor: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28%27%252F6dex5%21%27%25409-import3%2520from%2520%2522.0%2522B1%252F*7%25209art*%252F1.6toA1%2520%2520%2520%2520-9-7C3B%29%27%7E0%21%27.ns2ns8%253A4al%257B.4%257D24al8%27%29%7Eselected%21%270%27%290%252F7651%255Cn2A%2520origC%25203%2520MixRoot4glob5.9.css6in7mix8%2520%2919stA%2520%28B%253B1C6%253A%2501CBA9876543210_) 
- [x] crash when mixin Root alias with custom element selector: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28%27EHAGPzAY5WOfT8causes%2520crash%2520whg%2520mTed%2520-%2520remove%2520tokT%2520NJXA%253A%253AIC%2845-ItW%27%7E%252FN9H%255BA%255DGEzOB06LSF-rjV1ZmT%2520LSF-IV26L5QB3ZICL5-ItQ7X.aY-st-mTin8AW%27%7EP%21%27.rjKrjtW7.IKItW%27%29%7Eselected%21%27E%27%2947JJ5SU-A6%255D%2520.N__a%25207%255Cn8%253A%25209.st.cssAMTRjtB4%2540rule%255BCt%2520D%253B7E%252FU9FextgdGkrom%2520%2522.H%21%27%2540st-imporCIparJ%2520%2520KC%284Sq-L%28%2520NgtryO7%252F*%2520P%252Fq9Q%253B%2520%29Sid8TixUgrichVC%29%2520BWD%29X*%252F7Y%2520%284Z6.F__genjook%2520fqbasez%2522D%2501zqkjgZYXWVUTSQPONLKJIHGFEDCBA987654_)
- [ ] `@keyframes` are copied with root mixin and namespaced to mixin - should either namespace according to mixed-in sheet or reference the mixin origin: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28B%21%27%2540st-import%2520D%2520from%2520%2522.F%2522C000002%252F*8%257BreferenceG%2520name%257D%2520.entry__a9G7H%253B%2520%298%255B1%255D%257Bmixed-in%25206%257D%2520H2*%252F2.a920-st-47DC%29%27%7EF%21%27.root92057AC%2926%2520A9%29%27%29%7Eselected%21B%27%290EE2%255Cn3.st.css4mixin5animation6%2540keyframes7%253A%25208%25202%2540rule9%2520%28AjumpB%27%252Findex3C%253B2DMixRootE%2520%2520F%252F43G%25205H4__A%2501HGFEDCBA987654320_)
- [ ] JS mixin should namespace animation-name according to available symbols (from JS context? target context?): [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27fiYs%21%28%27%252FL.js%21%27moduY.exportS%253Dzunction%257B%257DARreturnAR4BP%255C%21%2520%255BRXBDBFXB8BFXBHBF4%255D6DM68M6vH%257DMF%295%29%27%7EJ%21%27%2540Timport%2520Zzrom%2520%2522.%252FL%2522%253BRXXQ%2540atruY%257Bstay%25207%257D%25208%2520qIv8%257DA%295QC%255B0G.O__aA9DE98E9HE5%29C%257BnSforK1GO__DNK2G8N%2520LW3GH5q.aAR-TLVZ%253B5%29%27%29%7EseYcted%21%27J%27%294UU5%255Cn6F4BI7global8conflict9RPVA%2520%28B%255C%27C5%2540ruYDunknownE%253B%2520%253FF%252CRG%255D%2520H7-nameI%2540keyframeSJ%252FO.st.cssK%2520sheetWLZinM%255C%21A%29NC%257B7%2520byOentryPanimationQ5%252F*%2520R54Ss%2520Tst-U%2520%2520V%253A%2520W%257D%255BX44YleZmixq*%252F5vT7%257Bz%2520f%2501zvqZYXWVUTSRQPONMLKJIHGFEDCBA987654_) 
- [x] when there are valid `-st-partial-mixin` and `-st-mixin`  with decls between them, the last mixin decl location is used as an insertion for both: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28E%21%21varsFa01%253B*b02O%298.MFxGa9*yGb9HAJCIDLHBJDICL8%29%27%29%7Eselected%21E%27%29*8KK0%253A%25206-Min0M%257Ba%252070---O%255Cn9%257D%253BCN63%252C%2520b%252049*DN-partial659*E%27%252Findex.st.cssF%2520%28*G0value%257BH8%2988.intoIbetweenJFbeforeK%2520%2520Lafter7MmixN7*-stO%253B8%2501ONMLKJIHGFEDC98760*_)
- [x] report and filter mixin of rules/atrules into `@keyframes`  states: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%287%21%27.9586%296.9Br5Fred%253B6%296%2540keyframes%2520my-name5%252F*%2520%2540rule%2520A5D82DbugC%255C%27%2526B%284red%29%255C%27%25209ed-in2%29%2520*%252F2A5D-st-9inC9%253B2%29210AE%296%29%27%29%7Eselected%217%27%2926D3%2520%25204color%253A5E26%255Cn7%27%252Fentry.st.css8Fgreen%253B9mixA0%2525B%253AhoveC%253A%2520D33E%2520%28F4%2520%2501FEDCBA98765432_)
- [ ] reported in [issue #2154](https://github.com/wix/stylable/issues/2154 ), mixed-in class with `-st-global` doesn't register the to `meta.globals`: [playground]()
- [x] reported in [issue #372](https://github.com/wix/stylable/issues/372 ), mixin of imported aliased element reports "unknown mixin": [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%284%21%270%2540st-import9from%2520%255C%27.5%255C%2770%252F*%2520remove%2520alias%2520to%2520fix%2520*%252F0.root368color%253A%2520aquamarine7%2900.local6%25208-st-mixin%253A37%29%27%7E5%21%270.big680%290%27%29%7Eselected%214%27%290%255Cn1%2520%25202.st.css9Comp4%27%252Findex25%252Fbutton26%2520%287%253B0801193%2520%25019876543210_)
- [x] reported in [issue #1203](https://github.com/wix/stylable/issues/1203 ), the JS mixin behavior differ on whether explicit or implicit parenthesis exist and on whether arguments are passed or not: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28T%21%27Gapp.z.csX%253AimZP-z-fromF%2522.Y%2522%253B3-z-namedFtwI%252C%2520nI%253B7%29UBEJJ34N5MEJJ34CE934%257B9LU%2520noB8nIN5M8nIC8nI%257B9L%27%7EY%21%27Gq.jXmodule.exZH%253DPtwWa%252C%2520b6Oa%257DKOb%257Dv%252C3nW6Kv7%29%27%29%7Eselected%21T%27%2937D4V2nd_undefined8twI5QFreZ%2520uslesHdiagnozicHJargument%2520at%25206%257DPDreturnPDDbackgroundF%2522green%25227%255Cn83*%252F3-z-qF9J%2520VfromRumentJBRX.no-parenthsiH%283%252F*3VnoQC%257BL7.partialPG3VnoQD%2520%2520E3V1z_F%253A%2520G%252F*%2520Hs%2520IoArgsJ%255C%27K%252C3DDcolorF%2522gold%2522L%257D%253B%25207%29Mindex%25200%2520iHemptyJN%253B7%297.no-argH%283G3-O%2520%252B%2520String%257BP%2520%283Q%2520reZR%2520argT%27%252Findex.z.cssU77Gq%2520withV-%2520WIFfunction%257BXH*%252F7Y%252Fjsmix.jsZport_R%2520%253D%2520qmixinv3D%29%253B3%29zst%2501zvq_ZYXWVUTRQPONMLKJIHGFEDCB9876543_)
- [x] local alias mix doesn't mix imported CSS: [playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%287%21%27%2540st-imporA%255B2%255D%2520from%2520%2522.B%252251.2.rooA%2891%252F*1expect6%2808%253B%2520-%2520noA2ed%2520bug%25209*%252F1.intoC144-st-8625%29%27%7EB%21%27.2C085%29%27%29%7Eselected%217%27%290144origin61%255Cn2mix3.st.css4%2520%25205%253B16%253A%25207%27%252Fentry382in90alias5%291At%2520B%252F83C%2520%28%2501CBA9876543210_)

## refactor
- processor
    - [x] `analyzeDeclaration` - move definition/validation to feature

- transform
    - [x] `transformLastPass` - move mixin append to feature hook
        
## deprecate
- [x] remove internal `SRule.mixins` usage

## test
- [x] move tests to `st-mixin.spec` and convert to new format
- [x] add missing tests

### CODE REVIEW

File: packages/core/src/features/st-import.ts [hunk 1]
```diff
@@ -113,7 +119,7 @@ export const hooks = createFeature<{
         if (rule.selector !== `:import`) {
             context.diagnostics.warn(
                 rule,
-                diagnostics.FORBIDDEN_DEF_IN_COMPLEX_SELECTOR(rootValueMapping.import)
+                diagnostics.FORBIDDEN_DEF_IN_COMPLEX_SELECTOR(PseudoImport)
```
Comment (barak007) [inline packages/core/src/features/st-import.ts:122]: can be inlined

File: packages/core/src/features/st-import.ts [hunk 2]
```diff
@@ -201,7 +207,7 @@ function validateImports(context: FeatureTransformContext) {
             const fromDecl =
                 importObj.rule.nodes &&
                 importObj.rule.nodes.find(
-                    (decl) => decl.type === 'decl' && decl.prop === valueMapping.from
+                    (decl) => decl.type === 'decl' && decl.prop === PseudoImportDecl.FROM
```
Comment (barak007) [inline packages/core/src/features/st-import.ts:210]: not sure we have benefit from putting these directives in consts

File: packages/core/src/stylable-meta.ts [hunk 3]
```diff
@@ -81,8 +84,7 @@ export class StylableMeta {
         const rootSymbol = CSSClass.addClass(context, RESERVED_ROOT_NAME);
         rootSymbol[valueMapping.root] = true;
 
-        setFieldForDeprecation(this, `mixins`, { objectType: `stylableMeta` });
-        this.mixins = [];
+        // setFieldForDeprecation(this, `mixins`, { objectType: `stylableMeta` });
```
Comment (barak007) [inline packages/core/src/stylable-meta.ts:87]: remove

File: packages/core/src/stylable-transformer.ts [hunk 4]
```diff
@@ -258,17 +257,18 @@ export class StylableTransformer {
...
-                this,
-                rule as SRule,
-                meta,
-                tsVarOverride || {},
-                cssVarsMapping,
-                path
-            )
-        );
+
+        const lastPathParams = {
```
Comment (barak007) [inline packages/core/src/stylable-transformer.ts:261]: lastPass

File: packages/core/src/stylable-utils.ts [hunk 5]
```diff
@@ -42,8 +42,18 @@ export function transformMatchesOnRule(rule: postcss.Rule, lineBreak: boolean) {
     return replaceRuleSelector(rule, { lineBreak });
 }
 
-export function mergeRules(mixinAst: postcss.Root, rule: postcss.Rule) {
+export const INVALID_MERGE_OF = (mergeValue: string) => {
+    return `invalid merge of: \n"${mergeValue}"`;
+};
+// ToDo: move to helpers/mixin
```
Comment (barak007) [inline packages/core/src/stylable-utils.ts:48]: move? now? later?

File: packages/core/src/stylable-value-parsers.ts [hunk 6]
```diff
@@ -58,8 +45,8 @@ export const valueMapping = {
     root: '-st-root' as const,
     states: '-st-states' as const,
     extends: '-st-extends' as const,
-    mixin: '-st-mixin' as const,
-    partialMixin: '-st-partial-mixin' as const,
+    mixin: '-st-mixin' as const, // ToDo: change to STMixin.MixinType.ALL,
```
Comment (barak007) [inline packages/core/src/stylable-value-parsers.ts:48]: I think we can deprecate these or remove. 
remove only usages in [SE](https://github.com/wixplosives/stylable-editor/pull/793/files)

File: packages/core/src/stylable-value-parsers.ts [hunk 7]
```diff
@@ -133,53 +120,6 @@ export const SBTypesParsers = {
...
-        mixinNode: postcss.Declaration,
-        strategy: (type: string) => 'named' | 'args',
-        diagnostics?: Diagnostics
-    ) {
-        return SBTypesParsers['-st-mixin'](mixinNode, strategy, diagnostics).map((mixin) => {
-            mixin.partial = true;
-            return mixin;
-        });
-    },
+    '-st-mixin': parseStMixin,
```
Comment (barak007) [inline packages/core/src/stylable-value-parsers.ts:123]: these `SBTypesParsers` are not used outside of Stylable so we can eliminate them and the maintenance of them

File: packages/core/test/helpers/rule.spec.ts [hunk 8]
```diff
@@ -64,6 +69,10 @@ describe(`helpers/rule`, () => {
                 { selector: '& &.x:hover' },
                 { selector: '&.y.x' },
                 { selector: '&&' }, // TODO: check if possible
+                { selector: ':not(&) &' },
+                { selector: ':nth-child(5n - 1 of &)' },
+                { selector: ':nth-child(5n - 2 of &, &)' },
+                { selector: ':nth-child(5n - 3 of &, .x, &)' }, // ToDo: check if to remove unrelated nested selectors
```
Comment (barak007) [inline packages/core/test/helpers/rule.spec.ts:75]: ?

Comment (idoros) [reply packages/core/test/helpers/rule.spec.ts:75]: the result of the the input `:nth-child(5n - 3 of i, .x, i)` is transformed into `:nth-child(5n - 3 of &, .x, &) {}` - the nested `.x` selector should probably be removed. 

However, there are cases where you would want to keep the nested selector: `.mix:is(:mix-state)`. I haven't had time go through it properly and fix. so comment for now.

File: packages/core/src/features/st-mixin.ts [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+    },
+};
+
+// HOOKS
+
+export const hooks = createFeature({
+    analyzeDeclaration({ context, decl }) {
+        ignoreDeprecationWarn(() => {
+            const parentRule = decl.parent as SRule;
+            const prevMixins = ignoreDeprecationWarn(() => parentRule?.mixins || []);
```
Comment (barak007) [inline packages/core/src/features/st-mixin.ts:77]: remove wrap

File: packages/core/src/features/st-mixin.ts [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+        } else if (partials.length) {
+            mixins = isInPartial ? mixins : partials.concat(mixins);
+        } else if (nonPartials.length) {
+            mixins = isInPartial ? nonPartials.concat(mixins) : mixins;
+        }
+    }
+    return mixins;
+}
+
+interface ApplyMixinConfig {
```
Comment (barak007) [inline packages/core/src/features/st-mixin.ts:216]: change to `ApplyMixinContext`

File: packages/core/test/features/css-keyframes.spec.ts [hunk 11]
```diff
@@ -475,4 +475,164 @@ describe(`features/css-keyframes`, () => {
...
+                        )}
+                        @rule[0] 0% { color: green } 
+                        @rule[1] 100% { } 
+                    */
+                    0% { -st-mixin: mix; }
+                    100% {}
+                }
+            `);
+        });
+        it.skip(`should not mix  @keyframes`, () => {
```
Comment (barak007) [inline packages/core/test/features/css-keyframes.spec.ts:615]: why skip?

Comment (idoros) [reply packages/core/test/features/css-keyframes.spec.ts:615]: leftover test - removed

**Pull request merged**